server/tests/test_migrations.py

```diff
@@ -0,0 +1,145 @@
+
+from parsec.components.postgresql import migrations as migrations_module
+from parsec.components.postgresql.handler import (
+    apply_migrations,
+    retrieve_migrations,
+)
+
+
+@pytest.fixture
+def pg_cluster(tmp_path) -> Iterator[Cluster]:
+    pg_cluster = TempCluster(data_dir_parent=tmp_path)
+    # Make the default superuser name stable.
+    pg_cluster.init(username="postgres")
+    pg_cluster.trust_local_connections()
+    pg_cluster.start(port="dynamic", server_settings={})
+    try:
+        yield pg_cluster
+    finally:
+        if pg_cluster.get_status() == "running":
+            pg_cluster.stop()
+        if pg_cluster.get_status() != "not-initialized":
+            pg_cluster.destroy()
+
+
+@pytest.fixture
+def pg_cluster_url(pg_cluster: Cluster) -> str:
+    spec = pg_cluster.get_connection_spec()
+    return f"postgresql://postgres@localhost:{spec['port']}/postgres"
+
+
+@pytest.fixture
+def pg_dump(pg_cluster: Cluster) -> str:
+    return pg_cluster._find_pg_binary("pg_dump")  # type: ignore[attr-defined]
+
+
+@pytest.fixture
+def pg_psql(pg_cluster: Cluster) -> str:
+    return pg_cluster._find_pg_binary("psql")  # type: ignore[attr-defined]
+
+
+@pytest.mark.postgresql
+@pytest.mark.asyncio
+async def test_migrations(pg_cluster_url, pg_psql, pg_dump):
+    async def execute_datamodel() -> None:
+        sql = importlib.resources.files(migrations_module).joinpath("datamodel.sql").read_text()
+        conn = await asyncpg.connect(pg_cluster_url)
+        try:
+            await conn.execute(sql)
+        finally:
+            await conn.close()
+
+    async def reset_db_schema() -> None:
+        conn = await asyncpg.connect(pg_cluster_url)
+        try:
+            rep = await conn.execute("DROP SCHEMA public CASCADE")
+            assert rep == "DROP SCHEMA"
+            rep = await conn.execute("CREATE SCHEMA public")
+            assert rep == "CREATE SCHEMA"
+        finally:
+            await conn.close()
+
+    async def dump_schema() -> str:
+        cmd = [pg_dump, "--schema=public", "--schema-only", pg_cluster_url]
+        print(f"run: {' '.join(cmd)}")
+        process = await asyncio.create_subprocess_exec(*cmd, stdout=asyncio.subprocess.PIPE)
+        stdout, stderr = await process.communicate()
+        return stdout.decode()
+
+    async def dump_data() -> str:
+        cmd = [pg_dump, "--schema=public", "--data-only", pg_cluster_url]
+        print(f"run: {' '.join(cmd)}")
+        process = await asyncio.create_subprocess_exec(*cmd, stdout=asyncio.subprocess.PIPE)
+        stdout, stderr = await process.communicate()
+        return stdout.decode()
+
+    async def restore_data(data: str) -> None:
+        cmd = [pg_psql, "--no-psqlrc", "--set", "ON_ERROR_STOP=on", pg_cluster_url]
+        print(f"run: {' '.join(cmd)}")
+        process = await asyncio.create_subprocess_exec(*cmd, stdin=asyncio.subprocess.PIPE)
+        await process.communicate(data.encode())
```

Comment:
`psql` will fail on first error, but will `communicate` fail too? 
As far as I understand, the error code will be stored in `process.returncode`, shouldn't we check the value? (or even check the process `stderr`)

I'm asking because the test seem to rely on that to check if restoring data worked or not.

Reply:
05bf51a06892c1295ab5a4648050882b124bed09